### ISSUE TITLE
Element-wise division of CV_64F matrices outputs inf instead of 0

### ISSUE THREAD

**Issue by taketwo** [opencv/opencv#8413]
##### System information (version)

- OpenCV => 3.2
- Operating System / Platform => Ubuntu 16.04
- Compiler => GCC 5.4.0

##### Detailed description

The documentation of [`divide`](http://docs.opencv.org/master/d2/de8/group__core__array.html#ga1f96b569cac4c286642b34eff098138e) says that

> `dst(I) = saturate(src1(I) * scale / src2(I))`
> When `src2(I)` is zero, `dst(I)` will also be zero.

In other words, it should be safe to divide by zero. However, the behavior I observe with `CV_64F` matrices is not consistent and depends on the size of the matrices.

For small matrices (up to 3 elements) division outputs zeros, as expected. However, if the size of the matrices is 4 or more, division outputs `inf`s.

Other data types behave as expected.

##### Steps to reproduce

```.cpp
#include <iostream>
#include <opencv2/core.hpp>

int main(int argc, const char** argv)
{
  cv::Mat A = cv::Mat::ones(1, 4, CV_64FC1);
  cv::Mat B = cv::Mat::zeros(1, 4, CV_64FC1);
  std::cout << "A = " << A << std::endl;
  std::cout << "B = " << B << std::endl;
  std::cout << "A / B = " << cv::Mat(A / B) << std::endl;
  return 0;
}
```

Output is:

```
A = [1, 1, 1, 1]
B = [0, 0, 0, 0]
A / B = [inf, inf, inf, inf]
```

##### Additional info

```
$ cat /proc/cpuinfo
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 61
model name      : Intel(R) Core(TM) i7-5500U CPU @ 2.40GHz
stepping        : 4
microcode       : 0x24
cpu MHz         : 2401.312
cache size      : 4096 KB
physical id     : 0
siblings        : 4
core id         : 0
cpu cores       : 2
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 20
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dtsacpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avxf16c rdrand lahf_lm abm 3dnowprefetch epb intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap xsaveopt dtherm ida arat pln pts
bugs            :
bogomips        : 4788.77
clflush size    : 64
cache_alignment : 64
address sizes   : 39 bits physical, 48 bits virtual
power management:
```

**Comment by alalek** [comment]
Probably this is documentation bug, like this: #6764

**Comment by csukuangfj** [comment]
Actually,
>When src2(I) is zero, dst(I) will also be zero

applies only to the second formula in the documentation. The first formula returns infinity.

**Comment by taketwo** [comment]
> Probably this is documentation bug

I do not think so. The behavior for *all* other types is consistent, division returns zeros.

**Comment by alalek** [comment]
Anyway this doesn't make sense for floating-point numbers. We should not violate floating-point standard IEEE754.

BTW, http://stackoverflow.com/questions/14682005/why-does-division-by-zero-in-ieee754-standard-results-in-infinite-value

**Comment by csukuangfj** [comment]
```cpp
std::cout << "1 / B = " << cv::Mat(1/ B) << std::end;
```
will print all zeros, instead of infinity.

**Comment by csukuangfj** [comment]
I have checked that for `A/B`
 - `(n/4)*4` elements are inf,  
 - `n%4` elements are 0.
where n is the number of elements in the matrix.

Probably there is some bug in the SIMD implementation.

**Comment by taketwo** [comment]
Yes, returning zeros is against the standard. However:

1. In CV domain we often need "safe" division that is guaranteed to not produce `NaN`s or `Inf`s, because these are not valid entries in image arrays.
2. The current behavior is documented and is (mostly) consistent. All other types besides from 64-bit floats give zeros. And, as Fangjun pointed, scalar by matrix division (1.0 / B) also returns zeros.

Therefore, while changing the implementation to match IEEE standard is of questionable value for most applications, it would certainly be a breaking and painful change. I think we should rather view this certain case where `Inf`s are returned at the moment as a bug and just fix it.

**Comment by taketwo** [comment]
@csukuangfj I can not reproduce this dependency on modulo 4 that you discovered. For me every matrix with size larger than 3 gives `Inf`s.

Thanks for the `cv:::setUseOptimized(false)` hint. It does the trick for me and I get zeros. In the problem I am trying to solve this is exactly the desired behavior.

**Comment by alalek** [comment]
It is hard to specify (and implement in optimal way) "safe" division for floating point numbers (+0, -0, FLP_EPSILON, 1e30 / 1e-30 (fp32) => Inf).

For integer data types "safety" makes sense and can be well defined.

**Comment by csukuangfj** [comment]
The implementation sets the result to 0 directly when divided by 0. I do not know why it happens to be inf only for CV_64F.

**Comment by vpisarev** [comment]
This issue will likely be solved (in one way or another) by OpenCV 4.0. The proposed solution is to retain the behaviour for integer data (0/0 => 0, x/0 => 0) and to implement correct IEEE-754 behaviour for floating-point data (0/0 => nan, x/0 => +/-inf, depending on the signs of x and 0). Any comments?

**Comment by taketwo** [comment]
:+1: as long as the behavior is consistent between float/double and different array sizes, of course.

**Comment by alalek** [comment]
BTW, `Mat` with size 1x4 is interpreted as Scalar in many OpenCV functions.

**Comment by yuriy-kozlov** [comment]
I am running into this problem with OpenCV 4.3 with 64F arrays. Looks like the official documented behavior changed in OpenCV 4.0 to follow standard IEEE-754. I was relying on the documented behavior in 3.x to get 0s for division by 0. Here is my workaround to do this explicitly:
```
cv::divide(1.0, A, Ainv);
Ainv.setTo(0, (A==0));
```
Posting in case others run into this when switching to OpenCV 4, because the comments above led me to believe the bug was fixed to follow the previous behavior.
